### PR TITLE
Fix encryption/decryption of messages

### PR DESCRIPTION
**Pull request by funkyboy**

### CODE REVIEW

File: lib/src/main/java/io/ably/lib/util/Crypto.java [hunk 1]
```diff
@@ -201,6 +201,7 @@ private CBCCipher(CipherParams params) throws AblyException {
 				encryptCipher = Cipher.getInstance(transformation);
 				encryptCipher.init(Cipher.ENCRYPT_MODE, params.keySpec, params.ivSpec);
 				decryptCipher = Cipher.getInstance(transformation);
+				decryptCipher.init(Cipher.DECRYPT_MODE, params.keySpec, params.ivSpec);
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/util/Crypto.java]: Why initialise both modes here when you're only going to use one of them?

File: lib/src/main/java/io/ably/lib/util/Crypto.java [hunk 2]
```diff
@@ -220,7 +221,13 @@ private CBCCipher(CipherParams params) throws AblyException {
 			System.arraycopy(plaintext, 0, cipherIn, 0, plaintextLength);
 			System.arraycopy(pkcs5Padding[padding], 0, cipherIn, plaintextLength, padding);
 			System.arraycopy(getIv(), 0, ciphertext, 0, blockLength);
-			byte[] cipherOut = encryptCipher.update(cipherIn);
+			byte[] cipherOut = new byte[0];
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/util/Crypto.java]: You don't need to assign anything to `cipherOut` here; just declare it.

File: lib/src/main/java/io/ably/lib/util/Crypto.java [hunk 3]
```diff
@@ -252,8 +258,13 @@ public String getAlgorithm() {
 		 * @return
 		 */
 		private byte[] getIv() {
-			if(iv == null)
-				return encryptCipher.update(emptyBlock);
+			if(iv == null) {
+				try {
+					return encryptCipher.doFinal(emptyBlock);
+				} catch (IllegalBlockSizeException|BadPaddingException e) {
+					e.printStackTrace();
```
Comment (paddybyers) [inline lib/src/main/java/io/ably/lib/util/Crypto.java]: This should be a `Log.e(TAG, ...`

File: lib/src/main/java/io/ably/lib/util/Crypto.java (hunk 1 above)
Comment (funkyboy) [reply lib/src/main/java/io/ably/lib/util/Crypto.java]: @paddybyers done https://github.com/ably/ably-java/pull/388/commits/864268a24b5482fd26406dc68bc7f778456bcbe7

File: lib/src/main/java/io/ably/lib/util/Crypto.java (hunk 2 above)
Comment (funkyboy) [reply lib/src/main/java/io/ably/lib/util/Crypto.java]: @paddybyers done https://github.com/ably/ably-java/pull/388/commits/864268a24b5482fd26406dc68bc7f778456bcbe7

File: lib/src/main/java/io/ably/lib/util/Crypto.java (hunk 3 above)
Comment (funkyboy) [reply lib/src/main/java/io/ably/lib/util/Crypto.java]: @paddybyers done https://github.com/ably/ably-java/pull/388/commits/d323a4bc8a6ece011cc42eda606ee29364dfafb1

**Comment by funkyboy** [comment]
@paddybyers This fixed crypto tests in rest: https://travis-ci.org/ably/ably-java/jobs/375816305#L2664

But now these fail :( https://travis-ci.org/ably/ably-java/jobs/375816305#L2219
Working on it.

**Comment by funkyboy** [comment]
@paddybyers Oops. Gradle is cheating a bit. Rest crypto tests are reported as passing but when I run them in Intellij I hit [this exception](https://github.com/ably/ably-java/blob/a4c33adc9988e823ddbb71c45766042cfee1d5e0/lib/src/main/java/io/ably/lib/types/BaseMessage.java#L111) which is triggered [from here](https://github.com/ably/ably-java/blob/2af72fb88e0d5e0b47820c1c1cab6a2235f5df73/lib/src/main/java/io/ably/lib/util/Crypto.java#L238).

Sample:
```
(ERROR): io.ably.lib.util.Crypto: decrypt()
javax.crypto.BadPaddingException: Given final block not properly padded
	at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:989)
	at com.sun.crypto.provider.CipherCore.doFinal(CipherCore.java:845)
	at com.sun.crypto.provider.AESCipher.engineDoFinal(AESCipher.java:446)
	at javax.crypto.Cipher.doFinal(Cipher.java:2223)
	at io.ably.lib.util.Crypto$CBCCipher.decrypt(Crypto.java:242)
	at io.ably.lib.types.BaseMessage.decode(BaseMessage.java:104)
	at io.ably.lib.types.MessageSerializer$MessageBodyHandler.handleResponseBody(MessageSerializer.java:116)
	at io.ably.lib.types.MessageSerializer$MessageBodyHandler.handleResponseBody(MessageSerializer.java:101)
	at io.ably.lib.http.BasePaginatedQuery.handleResponse(BasePaginatedQuery.java:160)
	at io.ably.lib.http.BasePaginatedQuery.handleResponse(BasePaginatedQuery.java:26)
	at io.ably.lib.http.HttpCore.handleResponse(HttpCore.java:334)
	at io.ably.lib.http.HttpCore.httpExecute(HttpCore.java:244)
	at io.ably.lib.http.HttpCore.httpExecute(HttpCore.java:146)
	at io.ably.lib.http.HttpCore.httpExecuteWithRetry(HttpCore.java:74)
	at io.ably.lib.http.HttpScheduler$AsyncRequest.httpExecuteWithRetry(HttpScheduler.java:291)
	at io.ably.lib.http.HttpScheduler$AblyRequestWithFallback.run(HttpScheduler.java:187)
	at io.ably.lib.util.CurrentThreadExecutor.execute(CurrentThreadExecutor.java:14)
	at io.ably.lib.http.HttpScheduler.ablyHttpExecuteWithFallback(HttpScheduler.java:388)
	at io.ably.lib.http.HttpScheduler.exec(HttpScheduler.java:86)
	at io.ably.lib.http.BasePaginatedQuery$1.execute(BasePaginatedQuery.java:78)
	at io.ably.lib.http.Http$Request.sync(Http.java:28)
	at io.ably.lib.http.BasePaginatedQuery$ResultRequest.sync(BasePaginatedQuery.java:209)
	at io.ably.lib.rest.Channel.history(Channel.java:119)
	at io.ably.lib.test.rest.RestCryptoTest.crypto_publish_key_mismatch(RestCryptoTest.java:203)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
(ERROR): io.ably.lib.types.MessageSerializer: Unexpected exception: Given final block not properly padded
(ERROR): io.ably.lib.util.Crypto: decrypt()
```

One step ahead, two steps back :(

**Comment by funkyboy** [comment]
Uhm, [this](https://github.com/ably/ably-java/pull/388#issuecomment-387146038) happens also in `develop` 🤔

**Comment by QuintinWillison** [comment]
There's no linked issue, nor a description of the issue being solved. This pull request has been left to languish for a long time, and it is not ready to land in its current state.

**Pull request closed without merging**